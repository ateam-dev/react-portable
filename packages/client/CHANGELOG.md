# @react-portable/client

## 0.0.2

### Patch Changes

- [#3](https://github.com/ateam-dev/react-portable/pull/3) [`d073b81`](https://github.com/ateam-dev/react-portable/commit/d073b8196680285c85bf70a5484df4fda7c8dff3) Thanks [@aiji42](https://github.com/aiji42)! - first release!

## 0.0.1

### Patch Changes

- [#1](https://github.com/ateam-dev/react-portable/pull/1) [`9e7866c`](https://github.com/ateam-dev/react-portable/commit/9e7866c26b7ca994a5d32a177759f66bafe3800c) Thanks [@aiji42](https://github.com/aiji42)! - first release
