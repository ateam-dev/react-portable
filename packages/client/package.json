{
  "name": "@react-portable/client",
  "version": "0.5.0",
  "repository": "git@github.com:ateam-dev/react-portable.git",
  "license": "Apache-2.0",
  "main": "./dist/index.mjs",
  "types": "./dist/index.d.ts",
  "exports": {
    ".": {
      "import": "./dist/index.mjs",
      "require": "./dist/index.cjs",
      "default": "./dist/index.mjs",
      "types": "./dist/index.d.ts"
    },
    "./web-components": {
      "import": "./dist/web-components.mjs",
      "require": "./dist/web-components.cjs",
      "default": "./dist/web-components.mjs",
      "types": "./dist/web-components.d.ts"
    },
    "./browser": {
      "default": "./dist/browser.umd.js"
    },
    "./*": "./*"
  },
  "typesVersions": {
    "*": {
      "web-components": [
        "./dist/web-components.d.ts"
      ]
    }
  },
  "files": [
    "dist"
  ],
  "scripts": {
    "build:main": "unbuild",
    "build:browser": "vite build",
    "build": "run-s build:browser build:main",
    "watch:main": "nodemon --ext ts,tsx --watch src -x 'npm run build:main'",
    "watch:browser": "vite build --watch",
    "watch": "run-p watch:*",
    "preview": "vite preview --port 3003",
    "dev": "run-p watch preview",
    "test": "vitest run",
    "test:coverage": "vitest run --coverage",
    "typecheck": "tsc --noEmit"
  },
  "devDependencies": {
    "@vitest/coverage-v8": "^0.33.0",
    "jsdom": "^22.0.0",
    "msw": "^1.2.2",
    "nodemon": "^3.0.1",
    "npm-run-all": "^4.1.5",
    "unbuild": "^2.0.0",
    "vite": "^4.4.3",
    "vite-plugin-dts": "^3.2.0",
    "vitest": "^0.33.0"
  },
  "dependencies": {
    "writable-dom": "^1.0.3"
  },
  "peerDependencies": {
    "react": ">=18"
  },
  "unbuild": {
    "clean": false
  }
}
