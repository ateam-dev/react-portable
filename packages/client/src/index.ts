export {
  registerReactPortable,
  parseFragmentId,
  createFragmentId,
  CLASS_NAME_FOR_GATEWAY_CACHE,
  CUSTOM_HEADER_KEY_GATEWAY,
} from "./react-portable";
