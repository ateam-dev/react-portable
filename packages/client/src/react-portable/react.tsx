import { useEffect, useReducer, useRef, useId } from "react";
import { register } from "./register";

export const ReactPortable = ({
  entry,
  gateway,
}: {
  entry: string;
  gateway?: string;
}) => {
  const isMounted = useMounted();
  const id = useId();
  const ref = useRef("");
  if (!isMounted && typeof window !== "undefined") {
    ref.current = document.getElementById(id)?.innerText ?? "";
    register();
  }

  return (
    <react-portable
      gateway={gateway}
      entry={entry}
      id={id}
      dangerouslySetInnerHTML={{ __html: ref.current }}
    />
  );
};

const useMounted = () => {
  const [isMounted, mounted] = useReducer(() => true, false);
  useEffect(() => {
    mounted();
  }, []);

  return isMounted;
};
