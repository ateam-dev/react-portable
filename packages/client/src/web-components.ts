export { register } from "./rp-preview/register";
export { RpPreview } from "./rp-preview/element";
