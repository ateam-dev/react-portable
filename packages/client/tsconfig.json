{
  "compilerOptions": {
    "target": "ESNext",
    "lib": ["DOM", "DOM.Iterable", "ESNext"],
    "skipLibCheck": true,
    "strict": true,
    "module": "ESNext",
    "moduleResolution": "Node",
    "resolveJsonModule": true,
    "isolatedModules": true,
    "noEmit": true,
    "allowSyntheticDefaultImports": true
  },
  "include": ["src"],
}
