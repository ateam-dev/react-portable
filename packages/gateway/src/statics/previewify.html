<div id="pr-bar"></div>
<script src="https://unpkg.com/@babel/standalone/babel.min.js"></script>
<script type="text/babel" data-type="module">
  import React, { useReducer, useEffect } from "https://esm.sh/react@18.2.0";
  import ReactDOM from "https://esm.sh/react-dom@18.2.0";

  const initial = localStorage.getItem("previewify[active]") === 'true'

  const preview = () => {
    window.rpPreviewDispatchers?.forEach(([, dispatcher]) => {
      dispatcher();
    });
  };


  let socket
  const previewActivate = () => {
    socket = new WebSocket(
      `${window.location.protocol === "https:" ? "wss:" : "ws:"}//${
        window.location.host
      }/_ws`,
    );
    socket.addEventListener("open", preview)
    socket.addEventListener("message", (event) => {
      if (event.data === "reload") preview()
    });
    socket.addEventListener("error", (event) => {
      console.error("WebSocket Error:", event);
    });
    socket.addEventListener("close", () => {
      // try to re-connect
      setTimeout(previewActivate, 5000);
    });
    return () => socket?.close?.()
  };


  const PreviewifyBar = () => {
    const [isPreviewing, toggle] = useReducer((s) => {
      const newState = !s
      localStorage.setItem("previewify[active]", (newState).toString())
      if (newState) preview()
      return newState
    }, initial)

    useEffect(() => {
      if (isPreviewing) return previewActivate()
    }, [isPreviewing])

    return (
      <div className="rp-wrapper">
        <div className="rp-logo" style={isPreviewing ? { color: '#4bd865' } : undefined}>♻️ Previewify</div>
        <label className="rp-toggle-button">
          <input type="checkbox" defaultChecked={initial} onChange={() => toggle()} checked={isPreviewing} />
        </label>
      </div>
    );
  };
  ReactDOM.render(<PreviewifyBar />, document.getElementById('pr-bar'));
</script>
<style>
    .rp-wrapper {
        position: fixed;
        bottom: 0;
        left: 0;
        width: 100%;
        height: 50px;
        display: flex;
        justify-content: space-between;
        align-items: center;
        padding: 0 10px;
        background-color: rgba(0, 0, 0, 50%);
        backdrop-filter: blur(10px);
    }
    .rp-logo {
        font-weight: bold;
        color: white;
        font-family: ui-sans-serif;
        font-size: 16px;
    }
    .rp-toggle-button {
        display: inline-block;
        position: relative;
        width: 50px;
        height: 25px;
        border-radius: 50px;
        border: 3px solid #f3f3f3;
        box-sizing: content-box;
        cursor: pointer;
        transition: border-color .4s;
    }

    .rp-toggle-button:has(:checked) {
        border-color: #4bd865;
    }

    .rp-toggle-button::after {
        position: absolute;
        top: 50%;
        left: 2px;
        transform: translateY(-50%);
        width: 22.5px;
        height: 22.5px;
        border-radius: 50%;
        background-color: #f3f3f3;
        content: '';
        transition: left .4s;
    }

    .rp-toggle-button:has(:checked)::after {
        left: 26px;
        background-color: #4bd865;
    }

    .rp-toggle-button input {
        display: none;
    }
</style>