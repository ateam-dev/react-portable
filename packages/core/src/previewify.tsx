import React, {
  FunctionComponent,
  ComponentType,
  isValidElement,
  ReactNode,
  useEffect,
  useReducer,
  useRef,
  Dispatch,
  ReducerAction,
  ForwardRefExoticComponent,
  forwardRef,
} from "react";
import { RpPreview } from "@react-portable/client/web-components";

export interface PreviewifyComponent<T = {}> extends FunctionComponent<T> {
  __code: string;
  __forQwik: FunctionComponent<T>;
}

type InferProps<T> = T extends ComponentType<infer U>
  ? U
  : T extends ForwardRefExoticComponent<infer U>
  ? U
  : T extends (props: infer U) => JSX.Element
  ? U
  : never;

type Reducer = (
  s: { previewing: boolean; serial: string },
  a: "open" | "close" | undefined,
) => { previewing: boolean; serial: string };

export const previewify = <
  T extends
    | ForwardRefExoticComponent<any>
    | ComponentType
    | ((props: any) => JSX.Element),
>(
  Component: T,
  code: string,
  option: { props?: InferProps<T> } = {},
): PreviewifyComponent<InferProps<T>> => {
  const Wrapped = forwardRef<any, InferProps<T>>((props, ref) => {
    const [{ previewing, serial }, dispatcher] = useReducer<Reducer>(
      (_, a) => {
        return { previewing: a !== "close", serial: Date.now().toString(36) };
      },
      { previewing: false, serial: "" },
    );

    useEffect(() => {
      window.previewifyDispatchers.add(dispatcher);
      return () => {
        window.previewifyDispatchers.delete(dispatcher);
      };
    }, []);

    return previewing ? (
      <Previewify code={code} props={props} key={serial}>
        {/* @ts-ignore */}
        <Component ref={ref} {...props} />
      </Previewify>
    ) : (
      // @ts-ignore
      <Component ref={ref} {...props} />
    );
  }) as unknown as PreviewifyComponent<InferProps<T>>;

  const ForQwik = (props: InferProps<T>) => {
    const [isServer, onClient] = useReducer(() => false, true);
    useEffect(onClient, []);

    return (
      <Component
        {...qwikifyProps(option.props ?? props, isServer, !!option.props)}
      />
    );
  };

  Wrapped.__code = code;
  Wrapped.__forQwik = ForQwik;

  return Wrapped;
};

const Previewify = ({
  code,
  props,
  children,
}: {
  code: string;
  children: ReactNode;
  props: Record<string, unknown>;
}) => {
  const rpRef = useRef<RpPreview>(null);
  useEffect(() => {
    if (rpRef.current?.previewing) rpRef.current.rerender(props);
    else rpRef.current?.preview(props);
  }, [props]);

  return (
    <rp-preview code={code} ref={rpRef}>
      <rp-preview-area>{children}</rp-preview-area>
      <template>{rpOutlets(props).map((el) => el)}</template>
    </rp-preview>
  );
};

const rpOutlets = <T extends Record<string, unknown>>(
  props: T,
): JSX.Element[] => {
  return Object.entries(props).flatMap(([k, v]) => {
    const isChildren = (k: string, _: unknown): _ is ReactNode =>
      k === "children";
    if (
      isChildren(k, v) ||
      isValidElement(v) ||
      (Array.isArray(v) && v.length > 0 && v.every(isValidElement))
    ) {
      return <rp-outlet key={k} _key={k} children={v} />;
    }

    return [];
  });
};

const qwikifyProps = <T extends Record<string, unknown>>(
  props: T,
  ssr: boolean,
  overwritten: boolean,
): T => {
  return Object.fromEntries(
    Object.entries(props).map(([key, val]) => {
      if (overwritten) return [key, val];
      if (key === "children" || val === "__outlet__")
        return [key, ssr ? null : <rp-slot _key={key} />];

      if (typeof val !== "string") return [key, val];

      const [, uuid, name] = val.match(/__function__#(.*)#(.*)/) ?? [];
      if (uuid && name) {
        return [
          key,
          (...args: unknown[]) =>
            document
              .querySelector(`rp-preview[id="${uuid}"]`)
              ?.dispatchEvent(new CustomEvent(name, { detail: { key, args } })),
        ];
      }

      return [key, val];
    }),
  ) as T;
};

if (typeof window !== "undefined") {
  window.previewifyDispatchers ||= new Set<Dispatch<ReducerAction<Reducer>>>();
}

declare global {
  interface Window {
    previewifyDispatchers: Set<Dispatch<ReducerAction<Reducer>>>;
  }
}
