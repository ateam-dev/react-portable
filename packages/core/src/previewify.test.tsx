import {
  afterAll,
  afterEach,
  beforeAll,
  beforeEach,
  describe,
  expect,
  test,
  vi,
} from "vitest";
import { previewify } from "./previewify";
import { render, cleanup, act, fireEvent } from "@testing-library/react";
import { register } from "@react-portable/client/web-components";
import { MouseEventHandler, ReactNode } from "react";
import { rest } from "msw";
import { setupServer } from "msw/node";

const Sample = ({
  onClick,
  testId,
}: {
  onClick?: MouseEventHandler<HTMLDivElement>;
  testId?: string;
}) => (
  <div onClick={onClick} data-testid={testId}>
    sample
  </div>
);

const SampleWithChildren = ({
  children,
  element,
}: {
  children: ReactNode;
  element?: ReactNode;
}) => (
  <div>
    sample with children:{children} {element}
  </div>
);

const server = setupServer(
  rest.post(`https://basepage.com/_fragments/:code`, (req, res, ctx) => {
    return res(ctx.status(200), ctx.text(``));
  }),
);

const openPreview = () => {
  act(() =>
    window.previewifyDispatchers.forEach((dispatcher) => dispatcher("open")),
  );
  vi.advanceTimersByTime(500);
};

const closePreview = () => {
  act(() =>
    window.previewifyDispatchers.forEach((dispatcher) => dispatcher("close")),
  );
};

beforeAll(() => {
  window.__previewifyDebug = true;
  vi.useFakeTimers();
  server.listen({ onUnhandledRequest: "error" });
});
beforeEach(() => {
  cleanup();
  register();
});
afterEach(() => {
  server.resetHandlers();
});
afterAll(() => {
  server.close();
});

describe("previewify", () => {
  test("__code", () => {
    const Component = previewify(Sample, "foo");
    expect(Component.__code).toBe("foo");
  });

  describe("Component", () => {
    test("It is same as the original component when initial", () => {
      const Component = previewify(Sample, "foo");
      const wrapped = render(<Component />);
      const original = render(<Sample />);

      expect(wrapped.container.innerHTML).toBe(original.container.innerHTML);
    });

    test("Dispatchers are added to previewifyDispatchers when rendered and removed when unmounted", () => {
      const Component = previewify(Sample, "foo");
      const { unmount } = render(<Component />);
      render(<Component />);

      expect(window.previewifyDispatchers.size).toBe(2);

      unmount();

      expect(window.previewifyDispatchers.size).toBe(1);
    });

    test("It will be wrapped by <rp-preview /> when previewifyDispatchers will be called", () => {
      const Component = previewify(Sample, "foo");
      const { asFragment } = render(<Component />);

      openPreview();

      expect(asFragment()).toMatchSnapshot();

      // Close the preview to see the original component
      closePreview();
      expect(asFragment()).toMatchSnapshot();
    });

    test("Elements of type ReactElement will be also rendered in the template when previewifyDispatchers will be called", () => {
      const Component = previewify(SampleWithChildren, "foo");
      const { container } = render(
        <Component
          children={<div>children</div>}
          element={<div>element</div>}
        />,
      );

      openPreview();

      expect(container).toMatchSnapshot();
    });
  });

  describe("Component.__forQwik", () => {
    test("`__outlet__` is converted to <rp-slot /> and rendered", () => {
      const Component = previewify(SampleWithChildren, "foo");
      const { asFragment } = render(
        <Component.__forQwik children="__outlet__" element="__outlet__" />,
      );

      expect(asFragment()).toMatchSnapshot();
    });
    test("`__function__` is converted into a function that delegates the original handler", () => {
      const onClickSpy = vi.fn();
      const Component = previewify(Sample, "foo");
      const rendered = render(<Component onClick={onClickSpy} />);
      render(
        <Component.__forQwik
          // @ts-ignore
          onClick="__function__#dummy-uuid#rp-preview-event"
          testId="clickable"
        />,
      );

      openPreview();

      fireEvent.click(rendered.getByTestId("clickable"));

      expect(onClickSpy).toBeCalled();
    });
    test("Allow overriding props for previews", () => {
      const ComponentSample = previewify(Sample, "foo", {
        props: { testId: "overwritten-test-it" },
      });
      let rendered = render(<ComponentSample.__forQwik testId="test-id" />);

      expect(rendered.asFragment()).toMatchSnapshot();

      rendered.unmount();

      const ComponentSampleWithChildren = previewify(
        SampleWithChildren,
        "foo",
        {
          props: {
            children: <div>overwritten children</div>,
            element: <div>overwritten element</div>,
          },
        },
      );
      rendered = render(
        <ComponentSampleWithChildren.__forQwik
          children="__outlet__"
          element="__outlet__"
        />,
      );

      expect(rendered.asFragment()).toMatchSnapshot();
    });
  });
});
