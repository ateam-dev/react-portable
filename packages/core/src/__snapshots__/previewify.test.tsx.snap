// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`previewify > Component > Elements of type ReactElement will be also rendered in the template when rpPreviewDispatchers will be called 1`] = `
<div>
  <rp-preview
    code="foo"
    id="dummy-uuid"
  >
    <rp-preview-area>
      <div>
        sample with children:
        <div>
          children
        </div>
         
        <div>
          element
        </div>
      </div>
    </rp-preview-area>
    <template>
      <rp-outlet
        _key="children"
      >
        <div>
          children
        </div>
      </rp-outlet>
      <rp-outlet
        _key="element"
      >
        <div>
          element
        </div>
      </rp-outlet>
    </template>
  </rp-preview>
</div>
`;

exports[`previewify > Component > It will be wrapped by <rp-preview /> when rpPreviewDispatchers will be called 1`] = `
<DocumentFragment>
  <rp-preview
    code="foo"
    id="dummy-uuid"
  >
    <rp-preview-area>
      <div>
        sample
      </div>
    </rp-preview-area>
    <template />
  </rp-preview>
</DocumentFragment>
`;

exports[`previewify > Component.__forQwik > \`__outlet__\` is converted to <rp-slot /> and rendered 1`] = `
<DocumentFragment>
  <div>
    sample with children:
    <rp-slot
      _key="children"
    />
     
    <rp-slot
      _key="element"
    />
  </div>
</DocumentFragment>
`;
